aten/src/ATen/native/NNPACK.cpp

```diff
@@ -232,76 +236,75 @@ Tensor _nnpack_spatial_convolution(
         "Mismatched Tensor types in NNPack convolutionOutput");
   }
 
-  const size_t batch_size = input.size(0);
-  const size_t input_channels = input.size(1);
-  const size_t output_channels = weight.size(0);
-  const struct nnp_size input_size = {.width = (size_t)input.size(3),
-                                      .height = (size_t)input.size(2)};
-  const struct nnp_padding input_padding = {.top = (size_t)padding[0],
-                                            .right = (size_t)padding[1],
-                                            .bottom = (size_t)padding[0],
-                                            .left = (size_t)padding[1]};
-  const struct nnp_size kernel_size = {.width = (size_t)weight.size(3),
-                                       .height = (size_t)weight.size(2)};
-
-  // If we don't have a defined bias Tensor, we need to create one filled with
-  // zeroes
-  auto bias_ =
-      bias.defined() ? bias : at::zeros({weight.size(0)}, input.options());
+  const auto compute = [&](const size_t batch_size)-> nnp_status {
+    const auto algorithm = nnp_convolution_algorithm_auto;
+    const size_t input_channels = input.size(1);
+    const size_t output_channels = weight.size(0);
+    const struct nnp_size input_size = {
+        .width = (size_t)input.size(3),
+        .height = (size_t)input.size(2),
+    };
+    const struct nnp_padding input_padding = {
+        .top = (size_t)padding[0],
+        .right = (size_t)padding[1],
+        .bottom = (size_t)padding[0],
+        .left = (size_t)padding[1],
+    };
+    const struct nnp_size kernel_size = {
+        .width = (size_t)weight.size(3),
+        .height = (size_t)weight.size(2),
+    };
+    const struct nnp_size output_size = {
+        .width = (size_t)output.size(3),
+        .height = (size_t)output.size(2),
+    };
+    const nnp_size output_subsample = {
+        .width = stride[1],
+        .height = stride[0],
+    };
+
+    const auto input_ = input.contiguous();
+    // If we don't have a defined bias Tensor, we need to create one filled with zeroes
+    const auto bias_ = bias.defined() ? bias : at::zeros({weight.size(0)}, input.options());
+
+    const size_t input_size_per_batch = input_channels * input_size.width * input_size.height;
+    const size_t output_size_per_batch = output_channels * output_size.width * output_size.height;
+
+    for (size_t batch = 0u; batch < batch_size; ++batch) {
```

Comment:
I didn't trace into the details but looks like nnp_convolution_output() might be able to do better batching than for-loop, or it's equivalent?
Guess it's not a big deal for mobile use case as batch size is usually small (1)?

Reply:
Good point, I'll modify.